docs(how-to): add "use an existing private key" (#568)
ipfs/ipfs-docs: docs/how-to/use-existing-private-key.md

---
title: Use an existing private key
description: How to start an IPFS node with a pre-generated private key.
---

# Use an existing private key

There are two locations in which IPFS stores private keys:

> 1. In the config file (`~/.ipfs/config`), following the [config specification](https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#identityprivkey) to store the default private key.
> 2. In the Keystore (`~/.ipfs/keystore`), following the [fs-repo specification](https://github.com/ipfs/specs/blob/master/REPO_FS.md) to store all other private keys.

Author: ```suggestion
Location | Purpose | Specification
-------- | --------- | -------------
Config file (`~/.ipfs/config`) | Holds a single private key; used as the default for initializing the IPFS node, IPNS publishing, etc. | [config#identityprivkey](https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#identityprivkey)
Keystore (`~/.ipfs/keystore`) | Holds additional private keys the node has access to; can be used for IPNS signing, publishing, etc. | [fs-repo#keystore](https://github.com/ipfs/specs/blob/master/REPO_FS.md#keystore)
```

### Rendered:
Location | Purpose | Specification
-------- | --------- | -------------
Config file (`~/.ipfs/config`) | Holds a single private key; used as the default for initializing the IPFS node, IPNS publishing, etc. | [config#identityprivkey](https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#identityprivkey)
Keystore (`~/.ipfs/keystore`) | Holds additional private keys the node has access to; can be used for IPNS signing, publishing, etc. | [fs-repo#keystore](https://github.com/ipfs/specs/blob/master/REPO_FS.md#keystore)

@johnnymatthews, what do you think?

---
title: Use an existing private key
description: How to start an IPFS node with a pre-generated private key.
---

# Use an existing private key

> There are two locations in which IPFS stores private keys:

Author: From my understanding, IPFS stores at both locations:

Location | Purpose
-------- | ---------
Config file (`~/.ipfs/config`) | Only one private key is stored, used as the default/primary key when initializing the IPFS node and other actions where it seeks a default private key.
Keystore (`~/.ipfs/keystore`) | Multiple keys are stored, used for other actions (e.g., signing, publishing, etc.)

That's my understanding after researching the topic. I can take a second look to make sure and include references if you think it's necessary!

> There are two locations in which IPFS stores private keys:

Reviewer: Thanks for clarifying! References aren't necessary, but this table might be useful to add in somewhere!

1. In the config file (`~/.ipfs/config`), following the [config specification](https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#identityprivkey) to store the default private key.
2. In the Keystore (`~/.ipfs/keystore`), following the [fs-repo specification](https://github.com/ipfs/specs/blob/master/REPO_FS.md) to store all other private keys.

> Private keys are handled programmatically through code or via the CLI. Here we will show you how to use your existing private keys using both methods.

Author: Perfect! That's good to know, thank you!

> ## CLI (`go-ipfs`)

Reviewer: ```suggestion
## CLI

Use the CLI to manage your private keys.

### Go-IPFS
```

Splitting the page into _CLI_ and _Programmatically_ sections.

> There's no way of directly initializing an IPFS node using your private keys from IPFS's Go CLI at this moment. To learn more and look for possible workarounds, see the discussion on issue [ipfs/go-ipfs#4240](https://github.com/ipfs/go-ipfs/issues/4240).

Reviewer: ```suggestion
There's no way of directly initializing an IPFS node using your private keys from the Go-IPFS CLI at this time. To learn more and look for possible workarounds, see the discussion on issue [ipfs/go-ipfs#4240](https://github.com/ipfs/go-ipfs/issues/4240).
```

> You can, however, import private keys into the Keystore:

Reviewer: ```suggestion
However, you can import private keys into the keystore:
```

Author: Keystore here refers to a [specific feature of IPFS](https://github.com/ipfs/specs/blob/master/KEYSTORE.md). Shouldn't we keep that capitalized and treat it like a proper noun to keep users aware that we are referring to that particular IPFS feature and not some other generic keystore?

Reviewer: Ah, sorry I missed this one. I assumed we were talking about the generic _keystore_. In this case, we should use something like _IPFS keystore_:

```suggestion
You can, however, import private keys into the IPFS keystore:
```

```sh
ipfs key import CustomKeyName ./CustomKeyFile

> k51qzi5uqu5dlli90qkt7xcnnjsd6jkn5rbdg7scuv6leozp1ecokewk5rzbeh
```

> The content of the file holding the private key you want to import into the Keystore needs to be in the correct format to be valid.
> The format of Keystore files is outlined in the [fs-repo](https://github.com/ipfs/specs/blob/master/REPO_FS.md#keystore) specification.

Reviewer: ```suggestion
The content of the file holding the private key you want to import into the keystore needs to be in the correct format to be valid. The format of keystore files is outlined in the [fs-repo](https://github.com/ipfs/specs/blob/master/REPO_FS.md#keystore) specification.
```

Author: Same as [above](https://github.com/ipfs/ipfs-docs/pull/568#discussion_r545996569).

Author: Same as [above](https://github.com/ipfs/ipfs-docs/pull/568#discussion_r545996569).

Reviewer: ```suggestion
The content of the file holding the private key you want to import into the IPFS keystore needs to be in the correct format to be valid. The format of keystore files is outlined in the [fs-repo](https://github.com/ipfs/specs/blob/master/REPO_FS.md#keystore) specification.
```

> For more information on the CLI commands, you can check the help option for the specific command:
> 
> ```sh
> ipfs COMMAND --help
> ```

Reviewer: ```suggestion
```

Reviewer: We can assume that readers know how to access the CLI help.

> ## CLI - JS (`js-ipfs`)

Reviewer: ```suggestion
### JS-IPFS
```

> You can provide the private key inline as a string when initializing (see `--help` for info on valid formats):

Reviewer: ```suggestion
You can provide the private key inline as a string when initializing:
```

```sh
jsipfs init --private-key 'my-private-key'
```

> You can also import private keys into the Keystore:

Reviewer: ```suggestion
You can also import private keys into the keystore:
```

Author: Same as [above](https://github.com/ipfs/ipfs-docs/pull/568#discussion_r545996569).

Author: Same as [above](https://github.com/ipfs/ipfs-docs/pull/568#discussion_r545996569).

Reviewer: ```suggestion
You can also import private keys into the IPFS keystore:
```

As above, since this is referring to an IPFS specific keystore, we should point this out.

```sh
jsipfs key import CustomKeyName --passin='your-pem-password' --input=./CustomKeyFile

> imported QmcasS8sQuasoFb2MDXbmBDwatWdhbkXrmx7131Rban9GG CustomKeyName
```

> For more information on the CLI commands, you can check the help option for the specific command:
> 
> ```sh
> jsipfs COMMAND --help
> ```
> 

Reviewer: ```suggestion
```

> ## JavaScript

Reviewer: ```suggestion
## Programmatically

Manage your private keys programmatically.

### JavaScript
```

> Using pre-generated private keys with JavaScript is pretty straight-forward:

Reviewer: ```suggestion
Using pre-generated private keys with JavaScript is straight-forward:
```

1. Import the IPFS module.
2. Create the node, passing the desired private key in the [correct format](https://github.com/ipfs/js-ipfs/blob/master/docs/MODULE.md#optionsinit) to the `IPFS.create` function parameter:

```javascript
import IPFS from 'ipfs'

const myExistingPrivateKey = 'my-private-key'
const ipfs = await IPFS.create({ privateKey: myExistingPrivateKey })
```

> ## Go

Reviewer: ```suggestion
### Go
```

Unfortunately, there's no clear way of initializing an IPFS node using your private keys from IPFS's Go Library at this moment.

## Troubleshooting

> - Key not changing

Reviewer: ```suggestion
### Key not changing
```

  - If you already have an initialized repo (i.e., `~/.jsipfs` is not empty), the existing key from the config file (`~/.jsipfs/config`) will be used instead of the provided one ([ipfs/js-ipfs#2261](https://github.com/ipfs/js-ipfs/issues/2261#issuecomment-637449985)).

## Use cases

- Use the keys for IPNS publishing.
- Temporary and disposable keys.
- Use in CI/CD pipelines.
- Rotate the IPFS identity.
- Use keys created from a third-party.

> ## Security
> 
> You'll have to take care of the handling of keys and their security. Remember that private keys are your identity and the only way of seeing encrypted content, publishing as you, among other actions. It's a protection layer between the user and the raw data.
> 

Reviewer: ```suggestion
```

We can expect readers to understand the security implications of managing private keys already.
